### PR TITLE
replace2D API for MutableObservable2DArray

### PR DESCRIPTION
**Pull request by mtaweel**
Support fully nested 2D replace for `MutableObservable2DArray` using `Diff.swift` library's `.nestedExtendedDiff` API which works on 2D arrays to allow multi-sections arrays observing of series of events (`deleteSection`, `deleteItems`, `insertSections`, `insertItems`, `moveSection`, `moveItem`) that migrate the old `MutableObservable2DArray` to the new `MutableObservable2DArray`

### Sample:
```
let array2D = MutableObservable2DArray([
      Observable2DArraySection(metadata: "units", items: [1,2,3]),
      Observable2DArraySection(metadata: "tens", items: [10,20,30]),
      Observable2DArraySection(metadata: "hundreds", items: [100,200,300]),
      Observable2DArraySection(metadata: "thousands", items: [1000,2000,3000]),
    ])

let newArray = Observable2DArray([
      Observable2DArraySection(metadata: "tens", items: [10,30]),
      Observable2DArraySection(metadata: "hundreds", items: [100,200,400]),
      Observable2DArraySection(metadata: "millions", items: [1000000,2000000,3000000]),
      Observable2DArraySection(metadata: "units", items: [4,3,2]),
      ])

array2D.replace2D(with: newArray, performDiff: true)
```

### CODE REVIEW

File: Sources/Collections/Observable2DArray.swift [hunk 1]
```diff
@@ -27,17 +27,17 @@ import Diff
 
 public enum Observable2DArrayChange {
   case reset
-
+
```
Comment (srdanrasic) [inline Sources/Collections/Observable2DArray.swift]: Would you mind enabling these two options https://i.stack.imgur.com/KjGnz.png so we don't get the unnecessary changes?

File: Sources/Collections/Observable2DArray.swift [hunk 2]
```diff
@@ -359,6 +388,14 @@ extension MutableObservable2DArray {
     sections[index] = section
     subject.next(Observable2DArrayEvent(change: .updateSections([index]), source: self))
   }
+  
+  /// Replace the entier 2d array with a new one forcing a reload
+  public func replace2D(with list: Observable2DArray<SectionMetadata, Item>)  {
```
Comment (srdanrasic) [inline Sources/Collections/Observable2DArray.swift]: I think we should rename this to just `replace`. Replacing it with another `Observable2DArray` implies that we are doing 2d replace. 

We should also rename `list` to `array` because it's array.

File: Sources/Collections/Observable2DArray.swift [hunk 3]
```diff
@@ -408,3 +445,98 @@ extension MutableObservable2DArray where Item: Equatable {
...
 }
 
+extension MutableObservable2DArray where Item: Equatable, SectionMetadata: Equatable {
+  
+  /// Replace the entire 2DArray performing nested diff (if preformDiff is true) on all
+  /// sections and section's items resulting in a series of events (deleteSection,
+  /// deleteItems, insertSections, insertItems, moveSection, moveItem) that migrate the old
+  /// 2DArray to the new 2DArray
+  /// Note that both Item and SectionMetadata should be Equatable
+  public func replace2D(with list: Observable2DArray<SectionMetadata, Item>, performDiff: Bool) {
```
Comment (srdanrasic) [inline Sources/Collections/Observable2DArray.swift]: Same comment as previous :)

File: Sources/Collections/Observable2DArray.swift [hunk 4]
```diff
@@ -408,3 +445,98 @@ extension MutableObservable2DArray where Item: Equatable {
...
+      
+      subject.next(Observable2DArrayEvent(change: .endBatchEditing, source: self))
+      lock.unlock()
+    } else {
+      replace2D(with: list)
+    }
+  }
+}
+
+struct NestedBatchUpdate {
```
Comment (srdanrasic) [inline Sources/Collections/Observable2DArray.swift]: I guess this could be private.

**Comment by mtaweel** [comment]
Thanks @srdanrasic. I applied your comments to the pull request

**Comment by srdanrasic** [comment]
Great, thanks a lot @mtaweel!
